### PR TITLE
State: Bump the console dispatcher to the front

### PR DESCRIPTION
**Pull request by blowery**
This bumps the console dispatcher to the front of the store dispatcher queue. This allows dispatches from the console dispatcher to be seen by the other store enhancers, like the data middleware.

To test, try dispatching a redux action from the console and catch it in the data middleware. 

Rest of the application should still work as normal.

### CODE REVIEW

File: client/state/index.js [hunk 1]
```diff
@@ -128,7 +128,8 @@ export function createReduxStore( initialState = {} ) {
 		enhancers.push( sitesSync );
 
 		if ( window.app && window.app.isDebug ) {
-			enhancers.push( consoleDispatcher );
+			// this has to be the first enhancer in the chain for the other enhancers to see it
+			enhancers.unshift( consoleDispatcher );
```
Comment (dmsnell) [inline client/state/index.js]: One way I was planning on cleaning this up is to stop relying on these imperative semantics and refocus the middleware chain in the declarative way it should be.

For example, this change fixes the issue at hand, but doesn't solve the problem that it's unclear which enhancers come in which order.

What if instead we could see a list in sequence identifying enhancers and the conditions in which they come in?

```js
const always = () => true;
const first = ( [ a, b ] ) => a;
const identity = a => a;
const whenInWindow = () => 'object' === typeof window;

const enhancers = [
	[ consoleDispatcher, whenInWindow ],
	[ devToolsExtension, () => whenInWindow() && window.devToolsExtension ],
	[ applyMiddleware( ...middleware ), always ],
	[ sitesSync, whenInWindow ],
].filter( identity ).map( first );
```

or differently

```js
const inWindow = 'object' === typeof window;

const enhancers = [
	inWindow && consoleDispatcher,
	inWindow && window.devToolsExtension && window.devToolsExtension(),
	applyMiddleware( ...middleware ),
	inWindow && sitesSync,
].filter( identity );
```

different approaches but each having the goal of trying to represent the final sequence of enhancers clearly and also the rules which dictate when they will be there.

Comment (blowery) [inline client/state/index.js]: How does that first filter invocation work?

Comment (dmsnell) [inline client/state/index.js]: > How does that first filter invocation work?

oops. it doesn't. :smile: this is why I tell people to never copy and paste my code from review comments. it was stream-of-thought coding.

```js
const enhancers = [
	…
].filter( ( [ first, second ] ) => !! second ).map( first )
```

alternatively as a reducer

```js
const enhancers = [
	…
].reduce( ( [ finalChain, [ enhancer, shouldInclude ] ) =>
	shouldInclude
		? finalChain.concat( enhancer )
		: finalChain
, [] );
```

**Comment by blowery** [comment]
@dmsnell fixed up

**Comment by dmsnell** [comment]
looks so readable @blowery!

are we sure that the ordering is correct? the fact that the console dispatcher comes _first_ while the Redux dev tools come _last_ makes me nervous.

**Comment by blowery** [comment]
I think that's because one is intercepting dispatches and one is generating them? Though I guess you can generate dispatches from the dev tools too... Hrm.

Maybe the data-layer store enhancer just needs to come last?

**Comment by dmsnell** [comment]
> Maybe the data-layer store enhancer just needs to come last?

I'm confused. Did you mean to write _the console dispatcher_?

**Comment by blowery** [comment]
Nope. When I had the console dispatcher _after_ the data layer enhancer, the data layer enhancer never saw dispatches from the console dispatcher. I had to move the console dispatcher earlier in the chain for the data layer to see dispatches from it.

**Comment by blowery** [comment]
So it seems the console enhancer has to come before the data layer enhancer. I'm not sure what it means for the relative position of other things.

The question I still havn't answered is _why_.

**Comment by dmsnell** [comment]
> So it seems the console enhancer has to come before the data layer enhancer. I'm not sure what it means for the relative position of other things.
> 
> The question I still havn't answered is why.

This could be because of which `dispatch` actually gets passed around.

The enhancer literally replaces the dispatch function and hides the old one. If the console dispatcher gets injected after the data layer then it won't see the actions dispatched because they pass through the previous dispatch function but not through the wrapped one.

Do the Redux tools need to be last? My guess is that they would have to go next to the console tools. If not, then we probably have a bug in my console dispatcher that should be fixed.

Thanks for continuing to improve this part of the core app state!

**Comment by dmsnell** [comment]
enhancerchain

Looks good to me @blowery - :shipit:

**Pull request merged**